hail/src/test/scala/is/hail/expr/ir/agg/StagedBlockLinkedListSuite.scala

```diff
@@ -0,0 +1,225 @@
+
+class StagedBlockLinkedListSuite extends TestNGSuite {
+
+  def compile[T: TypeInfo](emit: EmitRegion => Code[T]): (Region => T) = {
+    val fb = EmitFunctionBuilder[Region, T]
+    val mb = fb.apply_method
+    mb.emit(emit(EmitRegion.default(mb)))
+    val f = fb.result(None)()
+    f(_)
+  }
+
+  def assertEvalsTo[T: TypeInfo](expected: T)(emit: EmitRegion => Code[T]) {
+    val f = compile(emit)
+    val result = Region.scoped(f)
+    assertEquals(result, expected)
+  }
+
+  def assertEvalsToRV(expected: Any, typ: PType)(emit: EmitRegion => Code[Long]) {
+    assert(typ.virtualType.typeCheck(expected))
+    val f: Region => Long = compile(emit)
+    val result = Region.scoped { region =>
+      val off = f(region)
+      SafeRow.read(typ, region, off)
+    }
+    assertEquals(result, expected)
+  }
+
+  def assertBLLContents(elemType: PType, expected: IndexedSeq[Any])(
+    emit: (StagedBlockLinkedList, EmitRegion) => Code[Unit]
+  ) {
+    assertEvalsToRV(expected, PArray(elemType)) { er =>
+      val bll = new StagedBlockLinkedList(elemType, er.mb)
+      Code(
+        bll.init(er.region),
+        emit(bll, er),
+        bll.toArray(er))
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////////
+
+  def numBlocks(er: EmitRegion, bll: StagedBlockLinkedList): Code[Int] = {
+    val count = er.mb.newField[Int]
+    Code(
+      count := 0,
+      bll.foreachNode { b => count := count + 1 },
+      count)
+  }
+
+  def sumElements(er: EmitRegion, bll: StagedBlockLinkedList): Code[Int] = {
+    assert(bll.elemType.isOfType(PInt32()))
+    val acc = er.mb.newField[Int]
+    Code(
+      acc := 0,
+      bll.foreachElemAddress { e =>
+        e.m.mux(Code._empty,
+          acc := acc + er.region.loadInt(e.value))
+      },
+      acc)
+  }
+
+  def some(v: Code[_]): EmitTriplet =
+    EmitTriplet(Code._empty, false, v)
+  def none: EmitTriplet =
+    EmitTriplet(Code._empty, true, Code._fatal("unreachable"))
+
+  //
+
+  @Test def testInitialBlock() {
+    assertEvalsTo(1) { case er@EmitRegion(mb, r) =>
+      val bll = new StagedBlockLinkedList(PInt32(), mb)
+      Code(bll.init(r), numBlocks(er, bll))
+    }
+    assertEvalsTo(StagedBlockLinkedList.defaultBlockCap) { case EmitRegion(mb, r) =>
+      val bll = new StagedBlockLinkedList(PInt32(), mb)
+      Code(bll.init(r), bll.capacity(bll.firstNode))
+    }
+  }
+
+  @Test def testPushNewBlocks() {
+    assertEvalsTo(3) { case er@EmitRegion(mb, r) =>
+      val bll = new StagedBlockLinkedList(PInt32(), mb)
+      Code(
+        bll.init(r),
+        bll.pushNewBlockNode(r, 20),
+        bll.pushNewBlockNode(r, 30),
+        numBlocks(er, bll))
+    }
+    assertEvalsTo(55) { case EmitRegion(mb, r) =>
+      val bll = new StagedBlockLinkedList(PInt32(), mb)
+      Code(
+        bll.init(r),
+        bll.pushNewBlockNode(r, 55),
+        bll.capacity(bll.lastNode))
+    }
+  }
+
+  @Test def testPushInts() {
```

Comment:
A thing that I like to do when testing data structures is to just test against a much simpler data structure---in this case just a scala array/ArrayBuilder. You can build a wrapper around the linked list implementation that mimics the ArrayBuilder interface on a relatively simple dataset. You may still want the tests for the specific linked list functions that don't eventually get exposed, but doing that would let you do a pretty sizable test on the interface in a fairly legible way, without having to manually enumerate the expected results.